### PR TITLE
[ATTN][DECODE] Add `softmax_lse` output support to paged-decode FlashAttention on XPU.

### PR DESCRIPTION
**Pull request by baodii**
## Summary

- Return a float32 LSE tensor with shape
  `[num_heads_q, total_seqlen_q]` when `return_softmax_lse=True`.
- Use normalized partial outputs and one natural-log LSE per split for
  split-K decode, matching the CUDA FlashAttention reduction contract.
- Preserve the existing output-only behavior when
  `return_softmax_lse=False`.
- Fix split planning and row-statistics issues uncovered by the new accuracy
  coverage.

## Implementation

For a single split, the FMHA epilogue writes the final output and LSE directly.

For true split-K execution, each attention split now produces:

```text
O_i   = normalized partial output
LSE_i = log(sum(exp(scores_i)))
```

`ReduceSplitK` combines these values as:

```text
LSE_global = logsumexp_i(LSE_i)
weight_i   = exp(LSE_i - LSE_global)
O          = sum_i(weight_i * O_i)
```

The reducer computes each split weight once, stores the normalized weights in
SLM, and reuses them across output dimensions. This replaces the previous
`Oaccum + exp_sums + max_logits` representation with
`Oaccum + softmax_lse_accum`, removing one intermediate statistics buffer.

The change also:

- keeps the statistics row mapping separate from the output-copy row mapping,
  fixing packed-GQA attention-sink statistics;
- preserves host-planned live splits in `ReduceSplitK`;
- makes host split planning account for the sliding-window tile offset; and
- avoids unnecessary LSE logarithm work on the output-only direct path.

## API behavior

```python
out, softmax_lse = flash_attn_varlen_func(
    ...,
    return_softmax_lse=True,
)
```

Paged decode supports LSE output with global attention, sliding windows,
attention sinks, explicit split counts, and compact host split plans.

## Testing

Ran all FlashAttention tests selected by `-k decode`:

```text
1180 passed, 444 skipped, 3500 deselected
```

The skips are existing intentional cases for unsupported sink/window
combinations, invalid non-contiguous single-token inputs, split-plan parameter
combinations, device-memory limits, and MLA SLM limits.

Added reference-based accuracy coverage for:

- explicit single-split decode;
- requested split-K with short-sequence fallback;
- true split-K with even partitions;
- true split-K with an uneven final split and partial K tile;
- true split-K with 16-token cache blocks;
- LSE-enabled versus output-only parity;
- split-K versus direct single-split parity;
- packed-GQA sink row mapping; and
- global, sliding-window, sink, and compact-plan decode paths.

### CODE REVIEW

**Comment by yma11** [comment]
@YizhouZ FMHA will also need some changes like [code](https://github.com/vllm-project/vllm-xpu-kernels/pull/528/changes#diff-af7adacab18fb1f7735d40ef7b6a5596a20984fd0db65ddb75975ebc989bc87d), right? If so, will you provide a PR so that I will test end2end?

**Comment by YizhouZ** [comment]
> @YizhouZ FMHA will also need some changes like [code](https://github.com/vllm-project/vllm-xpu-kernels/pull/528/changes#diff-af7adacab18fb1f7735d40ef7b6a5596a20984fd0db65ddb75975ebc989bc87d), right? If so, will you provide a PR so that I will test end2end?

i think prefill part has already supported lse, is this PR not sufficient for your to test?

**Comment by yma11** [comment]
> > @YizhouZ FMHA will also need some changes like [code](https://github.com/vllm-project/vllm-xpu-kernels/pull/528/changes#diff-af7adacab18fb1f7735d40ef7b6a5596a20984fd0db65ddb75975ebc989bc87d), right? If so, will you provide a PR so that I will test end2end?
> 
> i think prefill part has already supported lse, is this PR not sufficient for your to test?

We need to remove assert against `is_paged` in FMHA and also need to add compile config for lse=True, we can do it on model `Qwen/Qwen3-235B-A22B` first.

**Comment by baodii** [comment]
> > > @YizhouZ FMHA will also need some changes like [code](https://github.com/vllm-project/vllm-xpu-kernels/pull/528/changes#diff-af7adacab18fb1f7735d40ef7b6a5596a20984fd0db65ddb75975ebc989bc87d), right? If so, will you provide a PR so that I will test end2end?
> > 
> > 
> > i think prefill part has already supported lse, is this PR not sufficient for your to test?
> 
> We need to remove assert against `is_paged` in FMHA and also need to add compile config for lse=True, we can do it on model `Qwen/Qwen3-235B-A22B` first.

@yma11 @YizhouZ Please review the latest commit which resolve the issue Yan raised.

**Comment by yma11** [comment]
> > > > @YizhouZ FMHA will also need some changes like [code](https://github.com/vllm-project/vllm-xpu-
> > We need to remove assert against `is_paged` in FMHA and also need to add compile config for lse=True, we can do it on model `Qwen/Qwen3-235B-A22B` first.
> 
> @yma11 @YizhouZ Please review the latest commit which resolve the issue Yan raised.

LGTM now. I will try.

**Comment by yma11** [comment]
@baodii  please add below configs and then the PR can be merged.
```
diff --git a/csrc/xpu/attn/kernel_configs/chunk_prefill_default.conf b/csrc/xpu/attn/kernel_configs/chunk_prefill_default.conf
index 4beff91..cc8e43f 100644
--- a/csrc/xpu/attn/kernel_configs/chunk_prefill_default.conf
+++ b/csrc/xpu/attn/kernel_configs/chunk_prefill_default.conf
@@ -73,3 +73,9 @@
 # Qwen/Qwen3-Next-80B-A3B-Instruct, Qwen/Qwen3-Next-80B-A3B-Thinking
 256,true,true,false,false,false
 256,true,true,false,false,true
+
+# --- zai-org/GLM-4.7-Flash ----------
+256,false,true,false,false,false
+256,false,true,false,false,true
+256,false,false,false,false,true
+256,false,false,false,false,false

```

**Comment by yma11** [comment]
DCP:
image
PCP:
image

**Review by yma11 — APPROVED** [review]

**Review by YizhouZ — APPROVED** [review]
